### PR TITLE
Carry over type casts in rewritten conditions to not disturb method disambiguation (#1022)

### LINKED ISSUE #589: Impossible to explicitly call a specific method with null argument.
An issue I opened over in the Groovy JIRA was just closed as not reproducible.

[GROOVY-7845 - Impossible to explicitly call a specific method with null argument.](https://issues.apache.org/jira/browse/GROOVY-7845)

This means that the problem must be located in Spock itself.
- Clone the repository https://github.com/huxi/bugs
- cd bugs/groovy/calls-with-null
- ./gradlew

The example class has three public static 'toStringSpecial' methods. It's impossible to call toStringSpecial(Set) or toStringSpecial(String) with a null argument.

This is possible in Java by casting null to the desired type.

It's currently impossible (AFAIK) to use Spock for testing such Java methods.

Environment:

```
------------------------------------------------------------
Gradle 2.13
------------------------------------------------------------

Build time:   2016-04-25 04:10:10 UTC
Build number: none
Revision:     3b427b1481e46232107303c90be7b05079b05b1c

Groovy:       2.4.4
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_92 (Oracle Corporation 25.92-b14)
OS:           Mac OS X 10.11.5 x86_64
```

### LINKED ISSUE #1022: Spock destroys method disambiguation completely
Given this in a Java class:

```java
    static abstract class Java1 {
        public Java1(String s) { System.out.println("String"); }
        public Java1(Pattern p) { System.out.println("Pattern"); }
    }

    static class Java2 extends Java1 {
        public Java2(String s) { super(s); }
        public Java2(Pattern p) { super(p); }
    }

    static class Java3 extends Java1 {
        public Java3(String s) { super((String) s); }
        public Java3(Pattern p) { super((Pattern) p); }
    }

    static class Java4 {
        public Java4(String s) { System.out.println("String"); }
        public Java4(Pattern p) { System.out.println("Pattern"); }
    }
```

and this in a Spock specification:
```groovy
    static class Groovy1 extends Java1 {
        Groovy1(String s) { super(s) }
        Groovy1(Pattern p) { super(p) }
    }

    static class Groovy2 extends Java1 {
        Groovy2(String s) { super(s as String) }
        Groovy2(Pattern p) { super(p as Pattern) }
    }

    static abstract class Groovy3 {
        Groovy3(String s) { println "String" }
        Groovy3(Pattern p) { println "Pattern" }
    }

    static class Groovy4 extends Groovy3 {
        Groovy4(String s) { super(s) }
        Groovy4(Pattern p) { super(p) }
    }

    static class Groovy5 extends Groovy3 {
        Groovy5(String s) { super(s as String) }
        Groovy5(Pattern p) { super(p as Pattern) }
    }

    static class Groovy6 {
        Groovy6(String s) { println "String" }
        Groovy6(Pattern p) { println "Pattern" }
    }
```

If you then do in the Java class
```java
    public static void main(String[] args) {
        new Java1((String) null) { };
        new Java2((String) null);
        new Java3((String) null);
        new Java4((String) null);
    }
```
all works well and four times `String` is printed.

Now we add to the Spock specification these tests:
```groovy
    def foo1() {
        expect:
            new Java1(null as String) { }
    }

    def foo2() {
        expect:
            new Java2(null as String)
    }

    def foo3() {
        expect:
            new Java3(null as String)
    }

    def foo4() {
        expect:
            new Java4(null as String)
    }

    def foo5() {
        expect:
            new Groovy1(null as String)
    }

    def foo6() {
        expect:
            new Groovy2(null as String)
    }

    def foo7() {
        expect:
            new Groovy3(null as String) { }
    }

    def foo8() {
        expect:
            new Groovy4(null as String)
    }

    def foo9() {
        expect:
            new Groovy5(null as String)
    }

    def foo10() {
        expect:
            new Groovy6(null as String)
    }
```
And we surprisingly find that all of them fail.
Some I actually would expect to fail.
`foo7` and `foo8` are expected to fail, because they even fail in a plain Groovy console.
But `foo9` and `foo10` should work fine for example.

### PR DESCRIPTION
**Pull request by Vampire**
Fixes #1022
Fixes #589

---
This change is [Reviewable](https://reviewable.io/reviews/spockframework/spock/1124)

### CODE REVIEW

File: spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java [hunk 1]
```diff
@@ -495,16 +495,24 @@ public void visitTupleExpression(TupleExpression expr) {
...
-        new VariableExpression(SpockNames.VALUE_RECORDER),
-        resources.getAstNodeCache().ValueRecorder_Record,
-        new ArgumentListExpression(
-            AstUtil.createDirectMethodCall(
-                new VariableExpression(SpockNames.VALUE_RECORDER),
-                resources.getAstNodeCache().ValueRecorder_StartRecordingValue,
-                new ArgumentListExpression(new ConstantExpression(recordCount++))
-            ),
-            expr));
+    MethodCallExpression result = AstUtil.createDirectMethodCall(
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java]: Does it make sense to create this, if we discard it in the case of a `CastExpression`? Also doesn't this mess up the `recordCount`?

File: spock-specs/src/test/groovy/org/spockframework/smoke/DisambiguationInConditions.groovy [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+import java.util.regex.Pattern
+
+import static DisambiguationInConditionsTestee.Java1
+import static DisambiguationInConditionsTestee.Java2
+import static DisambiguationInConditionsTestee.Java3
+import static DisambiguationInConditionsTestee.Java4
+import static DisambiguationInConditionsTestee.Verifier
+
+class DisambiguationInConditions extends Specification {
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/smoke/DisambiguationInConditions.groovy]: We probably need new tests in the `org.spockframework.smoke.condition.ConditionRendering` Spec or a related one

File: spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java [hunk 3]
```diff
@@ -495,16 +495,24 @@ public void visitTupleExpression(TupleExpression expr) {
...
+        AstUtil.createDirectMethodCall(
+          new VariableExpression(SpockNames.VALUE_RECORDER),
+          resources.getAstNodeCache().ValueRecorder_StartRecordingValue,
+          new ArgumentListExpression(new ConstantExpression(recordCount++))
+        ),
+        expr));
+
+    // maintain type casts to preserve method disambiguation
+    // for example when giving a casted null as argument
+    if (expr instanceof CastExpression) {
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java]: This code deals with casts in conditions, but in `DisambiguationInConditions` I only see casts in the `when` and not the `then`/`expect` blocks

File: spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java (hunk 1 above)
Comment (Vampire) [reply spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java]: We don't discard it, it is the second argument to the `CastExpression` constructor and thus casts the value returned from the method call. It copies the cast from the recorded expression "around" the record call.

File: spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java (hunk 3 above)
Comment (Vampire) [reply spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java]: Wtf, you are right.
Usually I verify that my tests fail without my changes.
I don't know what I f**** up there, I need to rewrite those tests, thanks for catching.

Comment (Vampire) [reply spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java]: Ah, I had them first similar to how I reported it with `System.out.println` and so on, had them failing, fixed them and then refactored (and thereby broke) the tests to properly verify the result.
The tests are fixed now.

File: spock-specs/src/test/groovy/org/spockframework/smoke/DisambiguationInConditions.groovy (hunk 2 above)
Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/smoke/DisambiguationInConditions.groovy]: You think so?
I didn't touch any rendering.

Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/smoke/DisambiguationInConditions.groovy]: This comment was based on my wrong assumption above, that the `result` was not used in the cast case.

**Pull request merged**